core/src/main/java/software/amazon/awssdk/core/SdkRequestOverrideConfig.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright 2010-2017 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.core;
+
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/**
+ * Base per-request override configuration for all SDK requests.
+ */
+public abstract class SdkRequestOverrideConfig {
+
+    private final Map<String, String> additionalHeaders;
+
+    private final Map<String, List<String>> additionalQueryParameters;
+
+    private final Duration clientExecutionTimeout;
+
+    private final String appendUserAgent;
+
+    private ReadLimitInfo readLimitInfo;
+
+    protected SdkRequestOverrideConfig(Builder<?> builder) {
+        this.additionalHeaders = builder.additionalHeaders();
+        this.additionalQueryParameters = builder.additionalQueryParameters();
+        this.clientExecutionTimeout = builder.clientExecutionTimeout();
+        this.appendUserAgent = builder.appendUserAgent();
+        this.readLimitInfo = builder.readLimitInfo();
+    }
+
+    public Optional<Map<String, String>> additionalHeaders() {
+        return Optional.ofNullable(additionalHeaders);
+    }
+
+    public Optional<Map<String, List<String>>> additionalQueryParameters() {
+        return Optional.ofNullable(additionalQueryParameters);
+    }
+
+    public Optional<Duration> clientExecutionTimeout() {
+        return Optional.ofNullable(clientExecutionTimeout);
+    }
+
+    public Optional<String> appendUserAgent() {
+        return Optional.ofNullable(appendUserAgent);
+    }
+
+    public Optional<ReadLimitInfo> readLimitInfo() {
+        return Optional.ofNullable(readLimitInfo);
+    }
+
+    public abstract Builder<? extends Builder> toBuilder();
+
+    public interface Builder<B extends Builder> {
+        Map<String, String> additionalHeaders();
+
+        B additionalHeader(String name, String value);
+
+        B additionalHeaders(Map<String, String> customHeaders);
+
+        Map<String, List<String>> additionalQueryParameters();
+
+        B additionalQueryParameter(String name, String value);
```

Comment:
Should this be `addAdditionalQueryParameter` to match the Special notes for collection types here: https://github.com/aws/aws-sdk-java-v2/wiki/Client-Configuration?

Reply:
+1!